caproto/server/conversion.py

```diff
@@ -230,182 +231,433 @@ def group_to_device(group):
+                # Allow this one through
+                ...
+            elif base_metadata[field_name] == field_info:
                 # Skip base attrs
                 continue
 
-        type_ = field_info.type
-        size = field_info.size
-        prompt = field_info.prompt
+        size = field_info.get('size', 0)
+        prompt = field_info['prompt']
 
         # alarm = parent.alarm
         kwargs = {}
 
+        if 'value' in field_info:
+            kwargs['value'] = field_info['value']
+
         if type_ == 'DBF_STRING' and size > 0:
             type_ = 'DBF_UCHAR'
             kwargs['max_length'] = size
+            kwargs['report_as_string'] = True
         elif size > 1:
             kwargs['max_length'] = size
 
         if type_ == 'DBF_MENU':
-            # note: ordered key assumption here (py3.6+)
-            kwargs['enum_strings'] = (f'menus.{field_info.menu}'
+            kwargs['enum_strings'] = (f'menus.{field_info["menu"]}'
+                                      '.get_string_tuple()')
+        elif type_ == 'DBF_DEVICE':
+            kwargs['enum_strings'] = (f'menus.dtyp_{record_type}'
                                       '.get_string_tuple()')
 
         if prompt:
             kwargs['doc'] = repr(prompt)
 
-        if field_info.special == 'SPC_NOMOD':
+        if field_info.get('special') == 'SPC_NOMOD':
             kwargs['read_only'] = True
 
-        type_class = type_info[type_]
+        type_class = DBD_TYPE_INFO[type_]
+        attr_name = get_attr_name_from_dbd_prompt(
+            record_type, field_name, prompt)
+        yield attr_name, type_class, kwargs, field_info
+
+
+def get_attr_name_from_dbd_prompt(record_type, field_name, prompt):
+    'Attribute name for fields: e.g., "Sim. Mode Scan" -> "sim_mode_scan"'
+    attr_name = prompt.lower()
+    # If there's a parenthesized part not at the beginning, remove it:
+    # e.g., "Velocity (EGU/s)" -> "Velocity"
+    if '(' in attr_name and not attr_name.startswith('('):
+        attr_name = attr_name.split('(')[0].strip()
+
+    # Replace bad characters with _
+    attr_name = re.sub('[^a-z_0-9]', '_', attr_name, flags=re.IGNORECASE)
+    # Replace multiple ___ -> single _
+    attr_name = re.sub('_+', '_', attr_name)
+    # Remove starting/ending _
+    attr_name = attr_name.strip('_') or field_name.lower()
+    if keyword.iskeyword(attr_name):
+        attr_name = f'{attr_name}_'
+    return _fix_attr_name(record_type, attr_name, field_name)
+
+
+def _fix_attr_name(record_type, attr_name, field_name):
+    'Apply any transformations specified in ATTR_RENAMES/REPLACES'
+    manual_fix = ATTR_FIXES.get(record_type, {}).get(field_name, None)
+    if manual_fix is not None:
+        return manual_fix
+
+    attr_name = ATTR_RENAMES.get(attr_name, attr_name)
+    for re_replace, repl_to in ATTR_REPLACES.items():
+        attr_name = re_replace.sub(repl_to, attr_name)
+    return attr_name
+
+
+LINKABLE = {
```

Comment:
I think I understand what this is for, but a comment would make me feel more sure.